### PR TITLE
Combine declaration and init removing overriding assignment Bug 572440

### PR DESCRIPTION
**Pull request by tsmaeder**
# What it does

Combines the initialization and declaration of variables when cleaning up overriding assignments given two conditions:

1. The assignment expression is not used
2. The initialization expression either is passive or the assignment immediately follows the declaration of the variable.

Both conditions intend to ensure that behaviour is preserved.

Fixes https://bugs.eclipse.org/bugs/show_bug.cgi?id=572440

# How to test
* Turn on "Remove overridden assignment" on the "Unnecessary Code" cleanup configuration page
* Run `Source->Cleanup" with code like this:
```java
// Keep this comment
boolean reassignedVar = true;
reassignedVar = "\\n".equals(File.pathSeparator);
return reassignedVar;
```

### CODE REVIEW

File: org.eclipse.jdt.ui.tests/ui/org/eclipse/jdt/ui/tests/quickfix/CleanUpTest.java
Comment (jjohnstn) [inline org.eclipse.jdt.ui.tests/ui/org/eclipse/jdt/ui/tests/quickfix/CleanUpTest.java:17463]: Why doesn't the cleanup change this to:

  System.out.println();
  long reassignedVar = System.currentTimeMillis();

The declaration has no non-passive call so it can simply be removed and moved to the true initialization.

File: org.eclipse.jdt.ui.tests/ui/org/eclipse/jdt/ui/tests/quickfix/CleanUpTest.java
Comment (jjohnstn) [inline org.eclipse.jdt.ui.tests/ui/org/eclipse/jdt/ui/tests/quickfix/CleanUpTest.java:17469]: Similarly, I would expect the code to end up:

System.out.println();
String reassignedVar = File.pathSeparator;

**Review by jjohnstn — CHANGES_REQUESTED** [review]

File: org.eclipse.jdt.ui.tests/ui/org/eclipse/jdt/ui/tests/quickfix/CleanUpTest.java
Comment (tsmaeder) [reply org.eclipse.jdt.ui.tests/ui/org/eclipse/jdt/ui/tests/quickfix/CleanUpTest.java:17463]: I thought about that, but didn't for two reasons:

1. The cleanup does not do that now, even though it would make sense, as well
2. Some folks don't like their variable declarations interspersed with the rest of the code. I don't want to mess with those folks.

**Comment by vogella** [comment]
Btw "Unnecessary code" tab has lots of options so as an idea it would be nice if these entries are listed in alpha numerical order instead of a what seems to be a random order.

Running this new cleanup on the eclipse.platform.resources plug-ins results in the following error:

java.lang.reflect.InvocationTargetException
	at org.eclipse.jface.operation.ModalContext.run(ModalContext.java:395)
	at org.eclipse.jface.wizard.WizardDialog.run(WizardDialog.java:1033)
	at org.eclipse.ltk.ui.refactoring.RefactoringWizard.createChange(RefactoringWizard.java:662)
	at org.eclipse.ltk.ui.refactoring.RefactoringWizard.computeUserInputSuccessorPage(RefactoringWizard.java:504)
	at org.eclipse.ltk.ui.refactoring.UserInputWizardPage.computeSuccessorPage(UserInputWizardPage.java:77)
	at org.eclipse.ltk.ui.refactoring.UserInputWizardPage.getNextPage(UserInputWizardPage.java:113)
	at org.eclipse.jdt.internal.ui.fix.CleanUpRefactoringWizard$CleanUpConfigurationPage.getNextPage(CleanUpRefactoringWizard.java:437)
	at org.eclipse.jface.wizard.WizardDialog.nextPressed(WizardDialog.java:910)
	at org.eclipse.jface.wizard.WizardDialog.buttonPressed(WizardDialog.java:468)
	at org.eclipse.jface.dialogs.Dialog.lambda$0(Dialog.java:619)
	at org.eclipse.swt.events.SelectionListener$1.widgetSelected(SelectionListener.java:84)
	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:252)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4261)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1066)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4078)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3655)
	at org.eclipse.jface.window.Window.runEventLoop(Window.java:823)
	at org.eclipse.jface.window.Window.open(Window.java:799)
	at org.eclipse.ltk.ui.refactoring.RefactoringWizardOpenOperation.lambda$0(RefactoringWizardOpenOperation.java:190)
	at org.eclipse.swt.custom.BusyIndicator.showWhile(BusyIndicator.java:74)
	at org.eclipse.ltk.ui.refactoring.RefactoringWizardOpenOperation.run(RefactoringWizardOpenOperation.java:209)
	at org.eclipse.ltk.ui.refactoring.RefactoringWizardOpenOperation.run(RefactoringWizardOpenOperation.java:126)
	at org.eclipse.jdt.internal.ui.refactoring.actions.RefactoringStarter.activate(RefactoringStarter.java:41)
	at org.eclipse.jdt.internal.corext.refactoring.RefactoringExecutionStarter.startCleanupRefactoring(RefactoringExecutionStarter.java:257)
	at org.eclipse.jdt.internal.ui.actions.AllCleanUpsAction.performRefactoring(AllCleanUpsAction.java:71)
	at org.eclipse.jdt.internal.ui.actions.CleanUpAction.runOnMultiple(CleanUpAction.java:209)
	at org.eclipse.jdt.internal.ui.actions.CleanUpAction.run(CleanUpAction.java:116)
	at org.eclipse.jdt.ui.actions.SelectionDispatchAction.dispatchRun(SelectionDispatchAction.java:274)
	at org.eclipse.jdt.ui.actions.SelectionDispatchAction.run(SelectionDispatchAction.java:252)
	at org.eclipse.jface.action.Action.runWithEvent(Action.java:474)
	at org.eclipse.jface.action.ActionContributionItem.handleWidgetSelection(ActionContributionItem.java:580)
	at org.eclipse.jface.action.ActionContributionItem.lambda$4(ActionContributionItem.java:414)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4261)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1066)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4078)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3655)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1155)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1046)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:644)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:551)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:156)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:152)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:136)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:402)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:659)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:596)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1467)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1440)
Caused by: java.lang.IllegalArgumentException: Variable binding and scope node should be provided
	at org.eclipse.jdt.internal.corext.dom.VarDefinitionsUsesVisitor.<init>(VarDefinitionsUsesVisitor.java:58)
	at org.eclipse.jdt.internal.ui.fix.OverriddenAssignmentCleanUp$1.visit(OverriddenAssignmentCleanUp.java:115)
	at org.eclipse.jdt.core.dom.VariableDeclarationStatement.accept0(VariableDeclarationStatement.java:247)
	at org.eclipse.jdt.core.dom.ASTNode.accept(ASTNode.java:3180)
	at org.eclipse.jdt.core.dom.ASTNode.acceptChildren(ASTNode.java:3251)
	at org.eclipse.jdt.core.dom.Block.accept0(Block.java:128)
	at org.eclipse.jdt.core.dom.ASTNode.accept(ASTNode.java:3180)
	at org.eclipse.jdt.core.dom.ASTNode.acceptChild(ASTNode.java:3228)
	at org.eclipse.jdt.core.dom.LambdaExpression.accept0(LambdaExpression.java:199)
	at org.eclipse.jdt.core.dom.ASTNode.accept(ASTNode.java:3180)
	at org.eclipse.jdt.core.dom.ASTNode.acceptChildren(ASTNode.java:3251)
	at org.eclipse.jdt.core.dom.MethodInvocation.accept0(MethodInvocation.java:228)
	at org.eclipse.jdt.core.dom.ASTNode.accept(ASTNode.java:3180)
	at org.eclipse.jdt.core.dom.ASTNode.acceptChild(ASTNode.java:3228)
	at org.eclipse.jdt.core.dom.ExpressionStatement.accept0(ExpressionStatement.java:136)
	at org.eclipse.jdt.core.dom.ASTNode.accept(ASTNode.java:3180)
	at org.eclipse.jdt.core.dom.ASTNode.acceptChildren(ASTNode.java:3251)
	at org.eclipse.jdt.core.dom.Block.accept0(Block.java:128)
	at org.eclipse.jdt.core.dom.ASTNode.accept(ASTNode.java:3180)
	at org.eclipse.jdt.core.dom.ASTNode.acceptChild(ASTNode.java:3228)
	at org.eclipse.jdt.core.dom.MethodDeclaration.accept0(MethodDeclaration.java:677)
	at org.eclipse.jdt.core.dom.ASTNode.accept(ASTNode.java:3180)
	at org.eclipse.jdt.core.dom.ASTNode.acceptChildren(ASTNode.java:3251)
	at org.eclipse.jdt.core.dom.TypeDeclaration.accept0(TypeDeclaration.java:499)
	at org.eclipse.jdt.core.dom.ASTNode.accept(ASTNode.java:3180)
	at org.eclipse.jdt.core.dom.ASTNode.acceptChildren(ASTNode.java:3251)
	at org.eclipse.jdt.core.dom.CompilationUnit.accept0(CompilationUnit.java:258)
	at org.eclipse.jdt.core.dom.ASTNode.accept(ASTNode.java:3180)
	at org.eclipse.jdt.internal.ui.fix.OverriddenAssignmentCleanUp.createFix(OverriddenAssignmentCleanUp.java:100)
	at org.eclipse.jdt.internal.ui.fix.AbstractMultiFix.createFix(AbstractMultiFix.java:49)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring.calculateChange(CleanUpRefactoring.java:772)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring$CleanUpASTRequestor.calculateSolutions(CleanUpRefactoring.java:301)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring$CleanUpASTRequestor.acceptAST(CleanUpRefactoring.java:279)
	at org.eclipse.jdt.core.dom.CompilationUnitResolver.resolve(CompilationUnitResolver.java:1008)
	at org.eclipse.jdt.core.dom.CompilationUnitResolver.resolve(CompilationUnitResolver.java:692)
	at org.eclipse.jdt.core.dom.ASTParser.createASTs(ASTParser.java:954)
	at org.eclipse.jdt.internal.corext.dom.ASTBatchParser.createASTs(ASTBatchParser.java:82)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring$CleanUpFixpointIterator.next(CleanUpRefactoring.java:399)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring.cleanUpProject(CleanUpRefactoring.java:682)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring.checkFinalConditions(CleanUpRefactoring.java:642)
	at org.eclipse.ltk.core.refactoring.CheckConditionsOperation.run(CheckConditionsOperation.java:86)
	at org.eclipse.ltk.core.refactoring.CreateChangeOperation.run(CreateChangeOperation.java:122)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2369)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2394)
	at org.eclipse.ltk.internal.ui.refactoring.WorkbenchRunnableAdapter.run(WorkbenchRunnableAdapter.java:89)
	at org.eclipse.jface.operation.ModalContext$ModalContextThread.run(ModalContext.java:122)
Root exception:
java.lang.IllegalArgumentException: Variable binding and scope node should be provided
	at org.eclipse.jdt.internal.corext.dom.VarDefinitionsUsesVisitor.<init>(VarDefinitionsUsesVisitor.java:58)
	at org.eclipse.jdt.internal.ui.fix.OverriddenAssignmentCleanUp$1.visit(OverriddenAssignmentCleanUp.java:115)
	at org.eclipse.jdt.core.dom.VariableDeclarationStatement.accept0(VariableDeclarationStatement.java:247)
	at org.eclipse.jdt.core.dom.ASTNode.accept(ASTNode.java:3180)
	at org.eclipse.jdt.core.dom.ASTNode.acceptChildren(ASTNode.java:3251)
	at org.eclipse.jdt.core.dom.Block.accept0(Block.java:128)
	at org.eclipse.jdt.core.dom.ASTNode.accept(ASTNode.java:3180)
	at org.eclipse.jdt.core.dom.ASTNode.acceptChild(ASTNode.java:3228)
	at org.eclipse.jdt.core.dom.LambdaExpression.accept0(LambdaExpression.java:199)
	at org.eclipse.jdt.core.dom.ASTNode.accept(ASTNode.java:3180)
	at org.eclipse.jdt.core.dom.ASTNode.acceptChildren(ASTNode.java:3251)
	at org.eclipse.jdt.core.dom.MethodInvocation.accept0(MethodInvocation.java:228)
	at org.eclipse.jdt.core.dom.ASTNode.accept(ASTNode.java:3180)
	at org.eclipse.jdt.core.dom.ASTNode.acceptChild(ASTNode.java:3228)
	at org.eclipse.jdt.core.dom.ExpressionStatement.accept0(ExpressionStatement.java:136)
	at org.eclipse.jdt.core.dom.ASTNode.accept(ASTNode.java:3180)
	at org.eclipse.jdt.core.dom.ASTNode.acceptChildren(ASTNode.java:3251)
	at org.eclipse.jdt.core.dom.Block.accept0(Block.java:128)
	at org.eclipse.jdt.core.dom.ASTNode.accept(ASTNode.java:3180)
	at org.eclipse.jdt.core.dom.ASTNode.acceptChild(ASTNode.java:3228)
	at org.eclipse.jdt.core.dom.MethodDeclaration.accept0(MethodDeclaration.java:677)
	at org.eclipse.jdt.core.dom.ASTNode.accept(ASTNode.java:3180)
	at org.eclipse.jdt.core.dom.ASTNode.acceptChildren(ASTNode.java:3251)
	at org.eclipse.jdt.core.dom.TypeDeclaration.accept0(TypeDeclaration.java:499)
	at org.eclipse.jdt.core.dom.ASTNode.accept(ASTNode.java:3180)
	at org.eclipse.jdt.core.dom.ASTNode.acceptChildren(ASTNode.java:3251)
	at org.eclipse.jdt.core.dom.CompilationUnit.accept0(CompilationUnit.java:258)
	at org.eclipse.jdt.core.dom.ASTNode.accept(ASTNode.java:3180)
	at org.eclipse.jdt.internal.ui.fix.OverriddenAssignmentCleanUp.createFix(OverriddenAssignmentCleanUp.java:100)
	at org.eclipse.jdt.internal.ui.fix.AbstractMultiFix.createFix(AbstractMultiFix.java:49)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring.calculateChange(CleanUpRefactoring.java:772)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring$CleanUpASTRequestor.calculateSolutions(CleanUpRefactoring.java:301)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring$CleanUpASTRequestor.acceptAST(CleanUpRefactoring.java:279)
	at org.eclipse.jdt.core.dom.CompilationUnitResolver.resolve(CompilationUnitResolver.java:1008)
	at org.eclipse.jdt.core.dom.CompilationUnitResolver.resolve(CompilationUnitResolver.java:692)
	at org.eclipse.jdt.core.dom.ASTParser.createASTs(ASTParser.java:954)
	at org.eclipse.jdt.internal.corext.dom.ASTBatchParser.createASTs(ASTBatchParser.java:82)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring$CleanUpFixpointIterator.next(CleanUpRefactoring.java:399)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring.cleanUpProject(CleanUpRefactoring.java:682)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring.checkFinalConditions(CleanUpRefactoring.java:642)
	at org.eclipse.ltk.core.refactoring.CheckConditionsOperation.run(CheckConditionsOperation.java:86)
	at org.eclipse.ltk.core.refactoring.CreateChangeOperation.run(CreateChangeOperation.java:122)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2369)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2394)
	at org.eclipse.ltk.internal.ui.refactoring.WorkbenchRunnableAdapter.run(WorkbenchRunnableAdapter.java:89)
	at org.eclipse.jface.operation.ModalContext$ModalContextThread.run(ModalContext.java:122)

File: org.eclipse.jdt.ui.tests/ui/org/eclipse/jdt/ui/tests/quickfix/CleanUpTest.java
Comment (jjohnstn) [reply org.eclipse.jdt.ui.tests/ui/org/eclipse/jdt/ui/tests/quickfix/CleanUpTest.java:17463]: I don't think option 2 qualifies as a reason for not doing a cleanup that was requested by the end-user.  If someone doesn't like a cleanup result, they can undo.  I think a reasonable solution would be to use the current behavior when there is no non-passive changes between declaration and initialization and do it the 2nd way when required.  Either way, you should not have a non-changing test in the middle without a name to suggest it is known to fail.

**Comment by tsmaeder** [comment]
@vogella  I'll have a look.

**Comment by tsmaeder** [comment]
> Running this new cleanup on the eclipse.platform.resources plug-ins results in the following error:
>
>java.lang.reflect.InvocationTargetException
>at org.eclipse.jface.operation.ModalContext.run(ModalContext.java:395)

Observations:

1. It's enough to run the cleanup on the org.eclipse.resources project
2. The error also occurs in master, so it's unrelated to my change
3. The problem occurs because a variable decl fragment has not binding in a lambda.

**Comment by tsmaeder** [comment]
@vogella the only time I can reproduce this is in lambdas inside methods that have compile errors. It seems reasonable, that the resolution of bindings could fail in these circumstances. Therefore, I would suggest that we just guard against the missing binding and don't cleanup those occurrences.

File: org.eclipse.jdt.ui.tests/ui/org/eclipse/jdt/ui/tests/quickfix/CleanUpTest.java
Comment (jjohnstn) [reply org.eclipse.jdt.ui.tests/ui/org/eclipse/jdt/ui/tests/quickfix/CleanUpTest.java:17463]: You can't make assumptions about the end-user's preferences unless it is obviously ridiculous.  There are end-users who don't care about the declaration being moved and will wonder why you didn't do the cleanup as it is perfectly good Java syntax.  There are users who won't like the choice of a variable name we choose for a new for loop, etc., etc..  That's their issue, not ours.  It is a "not" an unusable scenario as you mention.  The key thing is: "does the cleanup break the code or change the behaviour".  If the answer is no, then you should probably do it.  Anyway, I'm not going to argue with you.  If you don't want to do it and want to make the cleanup only change the first declaration, fine, but change the test name so it is obvious that the test doesn't do anything and either add comments in the test or have the title describe why not.

**Comment by vogella** [comment]
> Therefore, I would suggest that we just guard against the missing binding and don't cleanup those occurrences.

+1, ignoring files with compile errors sounds good to me.

File: org.eclipse.jdt.ui.tests/ui/org/eclipse/jdt/ui/tests/quickfix/CleanUpTest.java
Comment (tsmaeder) [reply org.eclipse.jdt.ui.tests/ui/org/eclipse/jdt/ui/tests/quickfix/CleanUpTest.java:17463]: I made it a preference. Since it's on by default, I think the behaviour is 99% what you suggested.

Comment (jjohnstn) [reply org.eclipse.jdt.ui.tests/ui/org/eclipse/jdt/ui/tests/quickfix/CleanUpTest.java:17463]: Thanks Thomas.  Regarding my earlier comments about a test that does not do the cleanup: a lot of cleanups break out these types of failure tests to a separate test that has multiple methods, etc.. that all have scenarios where the cleanup/quickfix is not applicable and usually have multiple methods with names to indicate why the cleanup fails (e.g. doNotChangeWhenVariablesUsedElsewhere()).  Then, the test simply verifies that no changes are made to the file so no "expected" string needed to test against.  It also is clear from the test that the cleanups don't apply in the scenarios.

Comment (tsmaeder) [reply org.eclipse.jdt.ui.tests/ui/org/eclipse/jdt/ui/tests/quickfix/CleanUpTest.java:17463]: But that's not what's happening: the cleanup still happens, but in a different way, right? I can separate the "do the move" case into a different test method from the "don't do the move" case if that helps.

Comment (jjohnstn) [reply org.eclipse.jdt.ui.tests/ui/org/eclipse/jdt/ui/tests/quickfix/CleanUpTest.java:17463]: No issue.  I was just clarifying my initial comment about the original test when there was no change being made.  I believe you had asked this at our meeting and so I clarified here.  Sorry for the confusion.

**Comment by tsmaeder** [comment]
@jjohnstn are there things you still need addressed? The change can't be merged until you change your vote to "approve".

**Comment by jjohnstn** [comment]
Sorry, I was waiting for confirmation that Lars' exception issue was resolved.  I guess that is done.  I'll try out the latest patch and hopefully merge today.

**Comment by jjohnstn** [comment]
Worked great.

**Comment by vogella** [comment]
Thanks @tsmaeder and @jjohnstn. In the past it was a good practise to apply the JDT cleanups on JDT itselft (dogfooding) to find additional issues. Expect PR from me for JDT using this new cleanup :-)

**Comment by vogella** [comment]
That reminded me about https://github.com/eclipse-jdt/eclipse.jdt.ui/issues/34

**Pull request merged**
